pkg/audit/verifier.go

```diff
@@ -74,48 +76,88 @@ func (verifier *Verifier) Verify(ctx context.Context, stripe *Stripe, skip map[s
 	shareSize := pointer.GetRemote().GetRedundancy().GetErasureShareSize()
 	bucketID := createBucketID(stripe.SegmentPath)
 
-	// TODO(kaloyan): CreateAuditOrderLimits checks overlay cache if nodes are online and won't return
-	// order limits for offline node. So we miss to record them as offline during the audit
+	var offlineNodes storj.NodeIDList
+	var failedNodes storj.NodeIDList
+	containedNodes := make(map[int]storj.NodeID)
+	sharesToAudit := make(map[int]Share)
+
 	orderLimits, err := verifier.orders.CreateAuditOrderLimits(ctx, verifier.auditor, bucketID, pointer, skip)
 	if err != nil {
 		return nil, err
 	}
 
+	offlineNodes = getOfflineNodes(stripe.Segment, orderLimits)
+
 	shares, nodes, err := verifier.DownloadShares(ctx, orderLimits, stripe.Index, shareSize)
 	if err != nil {
 		return nil, err
 	}
 
-	var offlineNodes storj.NodeIDList
-	var failedNodes storj.NodeIDList
-	containedNodes := make(map[int]storj.NodeID)
-	sharesToAudit := make(map[int]Share)
-
 	for pieceNum, share := range shares {
-		if share.Error != nil {
-			// TODO(kaloyan): we need to check the logic here if we correctly identify offline nodes from those that didn't respond.
-			if share.Error == context.DeadlineExceeded || !transport.Error.Has(share.Error) || ContainError.Has(share.Error) {
-				containedNodes[pieceNum] = nodes[pieceNum]
-			} else {
+		if share.Error == nil {
+			// no error -- share downloaded successfully
+			sharesToAudit[pieceNum] = share
+			continue
+		}
+		if transport.Error.Has(share.Error) {
+			if errs.IsFunc(share.Error, func(err error) bool {
+				return err == context.DeadlineExceeded
+			}) {
+				// dial timeout
 				offlineNodes = append(offlineNodes, nodes[pieceNum])
+				verifier.log.Debug("dial timeout (offline)", zap.String("Node ID", nodes[pieceNum].String()), zap.Error(share.Error))
+				continue
 			}
-		} else {
-			sharesToAudit[pieceNum] = share
+			if errs.IsFunc(share.Error, func(err error) bool {
+				return status.Code(err) == codes.Unknown
```

Comment:
it could be a little confusing that we know `codes.Unknown` means a dial failed, but other codes _are_ unknown (below). I guess this will just be more clear once moved into helper functions.

Reply:
`codes.Unknown` usually means the error is not an RPC error. And it is not only this in this check. This if-then-else is wrapped in a check for a transport error class. So the actual check for dial failed is: "an error that is wrapped by the transport error class, but is not an RPC error or `context.DeadlineExceeded` error".

I agree these checks are confusing, but this is what we can do as error analysis at the moment. This was also the reason to implement all those TestDownloadShare* tests. So we know what error to expect in the different use cases. And have a test to fail if we don't get the expected error anymore.